### PR TITLE
Add `thrust::universal_host_pinned_vector`

### PR DESCRIPTION
**Pull request by bernhardmgruber**
We have all the pieces in Thrust to provide a vector backed by pinned host memory. Only a few aliases were missing. Furthermore, the unit tests were already covering a vector with pinned memory.

This PR adds the missing aliases and extends a few tests. The main new alias is `thrust::universal_host_pinned_vector<T>` which uses CUDA pinned host memory (`cudaMallocHost`), and the native memory resource on other systems (as dictated by the existing `universal_host_pinned_memory_resource` on each system).

Fixes: #2485

### CODE REVIEW

File: libcudacxx/examples/concurrent_hash_table.cu [hunk 1]
```diff
@@ -13,27 +13,13 @@
...
-  thrust::system::cuda::detail::cuda_memory_resource<thrust::system::cuda::detail::cudaMallocManaged, cudaFree, void*>;
-
-template <typename T>
-using universal_allocator = thrust::mr::stateless_resource_allocator<T, universal_raw_memory_resource>;
-
-template <typename T>
-using universal_vector = thrust::device_vector<T, universal_allocator<T>>;
-
-} // namespace thrust
-
```
Comment (bernhardmgruber) [inline libcudacxx/examples/concurrent_hash_table.cu:36]: This is a drive-by fix, since this functionality has been upstreamed into Thrust IIUC.

File: thrust/testing/unittest/testframework.h [hunk 2]
```diff
@@ -388,37 +388,33 @@ class UnitTestDriver
   void VTEST##Universal()                                                                                               \
   {                                                                                                                     \
     VTEST<thrust::universal_vector<int>>();                                                                             \
-    VTEST<thrust::device_vector<                                                                                        \
-      int,                                                                                                              \
-      thrust::mr::stateless_resource_allocator<int, thrust::universal_host_pinned_memory_resource>>>();                 \
```
Comment (bernhardmgruber) [inline thrust/testing/unittest/testframework.h:393]: Here, our unit tests already cover cost pinned memory.

**Review by miscco — APPROVED** [review]

File: thrust/thrust/system/cpp/memory_resource.h [hunk 3]
```diff
@@ -62,6 +62,7 @@ using memory_resource = detail::native_resource;
  *  \p mr::new_delete_resource and tags it with \p cpp::universal_pointer.
  */
 using universal_memory_resource = detail::universal_native_resource;
+// FIXME(bgruber): comment below is wrong or alias should be to universal_memory_resource
 /*! An alias for \p cpp::universal_memory_resource. */
 using universal_host_pinned_memory_resource = detail::native_resource;
```
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply thrust/thrust/system/cpp/memory_resource.h:67]: Tracking here: #2659
